src/editor/input_center.cpp

```diff
@@ -0,0 +1,879 @@
+      set_object();
+      break;
+    default:
+      break;
+  }
+}
+
+Rectf
+EditorInputCenter::tile_drag_rect() {
+  Rectf result = drag_rect();
+  result.p1 = sp_to_tp(result.p1);
+  result.p2 = sp_to_tp(result.p2);
+  return result;
+}
+
+Rectf
+EditorInputCenter::selection_draw_rect() {
+  Rectf select = tile_drag_rect();
+  select.p2 += Vector(1, 1);
+  select.p1 = tile_screen_pos(select.p1);
+  select.p2 = tile_screen_pos(select.p2);
+  return select;
+}
+
+void
+EditorInputCenter::update_tile_selection() {
+  Rectf select = tile_drag_rect();
+  auto tiles = Editor::current()->tileselect.tiles.get();
+  TileMap* tilemap = dynamic_cast<TileMap*>(Editor::current()->layerselect.selected_tilemap);
+  if ( !tilemap ) {
+    return;
+  }
+
+  tiles->tiles.clear();
+  tiles->width = select.get_width() + 1;
+  tiles->height = select.get_height() + 1;
+
+  int w = tilemap->get_width();
+  int h = tilemap->get_height();
+  for (int y = select.p1.y; y <= select.p2.y; y++) {
+    for (int x = select.p1.x; x <= select.p2.x; x++) {
+      if ( x < 0 || y < 0 || x >= w || y >= h) {
+        tiles->tiles.push_back(0);
+      } else {
+        tiles->tiles.push_back(tilemap->get_tile_id(x, y));
+      }
+    }
+  }
+}
+
+void
+EditorInputCenter::event(SDL_Event& ev) {
+  auto tileselect = &(Editor::current()->tileselect);
+  switch (ev.type) {
+    case SDL_MOUSEBUTTONDOWN:
+    switch (ev.button.button) {
+      case SDL_BUTTON_LEFT: {
+        process_left_click();
+      } break;
+      case SDL_BUTTON_RIGHT: {
+        process_right_click();
+      } break;
+    } break;
+
+    case SDL_MOUSEBUTTONUP:
+      dragging = false;
+      break;
+
+    case SDL_MOUSEMOTION:
+    {
+      mouse_pos = VideoSystem::current()->get_renderer().to_logical(ev.motion.x, ev.motion.y);
+      actualize_pos();
+      if (dragging) {
+        switch (tileselect->input_type) {
+          case EditorInputGui::IP_TILE:
+            if (dragging_right) {
+              update_tile_selection();
+            } else {
+              switch (tileselect->select_mode->get_mode()) {
+                case 0:
+                  put_tile();
+                  break;
+                case 1:
+                  draw_rectangle();
+                  break;
+                default:
+                  break;
+              }
+            }
+            break;
+          case EditorInputGui::IP_OBJECT:
+            if (tileselect->object.empty()) {
+              if (tileselect->select_mode->get_mode() == 1 &&
+                  tileselect->object.empty() ) {
+                rubber_rect();
+              }
+            } else {
+              move_object();
+            }
+            break;
+          default:
+            break;
+        }
+      }
+    } break;
+    case SDL_KEYDOWN:
+      if (ev.key.keysym.sym == SDLK_F8) {
+        render_grid = !render_grid;
+      }
+      break;
+    default:
+      break;
+  }
+}
+
+void
+EditorInputCenter::actualize_pos() {
```

Comment:
This function should be called "update_pos". You're becoming more and more German every second :P (because "aktualisieren").